README.md
# BCLC Numbers

A Perl, [Dancer2](https://metacpan.org/pod/Dancer2)-based web application with a
front-end that allows a user to input the six numbers related to BCLC's Lotto
6/49 draws, and returns historical win/loss information back to 1982.

## Table of Contents

- [Installation](#installation)
- [Using the UI](#using-the-ui)
- [Options](#options)
- [Documentation](#documentation)
- [Test Results](#testing)

## Installation

- Download the application's [zip file](https://github.com/ddejarisco/codechallenge-REQ59847-00003/archive/master.zip)
- Unzip the archive
- Change into the extracted directory
- `perl Makefile.PL`
- `make`
- `make test`
- `make install`
- Install [Plack](https://metacpan.org/pod/Plack) using `cpanm Plack`
- Start the minimal web server: `plackup bin/app.pl`

## Using the UI

- Go to the web page [http://localhost:5000](http://localhost:5000) (replace `localhost` with your 
server's IP/DNS name if required)
- Type in six unique number into each one of the input text boxes. The border of
each box will turn green if the number is valid, and red if not. Warnings will
also be displayed dynamically to explain the specific problem
- The results are shown dynamically, immediately after we've collected six valid
numbers
- There is no submit button. For a new set of results, simply change one or more
of the previously entered numbers
- To reset all [Options](#options) listed on the page, as well as all of the
previously entered numbers, simply click the `Reset/Clear` button

## Options

There are a few options available on the page. Simply hover over the "Options"
in the top menu bar to access them.

#### Data Source

This option provides the user the ability to change between operating on either
the original CSV file of historical data, or a previously compiled SQLite 
database with the same CSV data.

##### Values

    CSV
    
Use the original CSV file as the data source. This is the default.

    DB
    
Use the previously compiled SQLite database for all operations.

#### Display UI Help

This checkbox toggles a basic list of help items on the screen dynamically.

#### Include wins less than $85

By default, we list in the results, wins that are $85 and up. Check this
checkbox to display all wins, regardless of the value.        

## Documentation

The various code files each have their own documentation in either markdown or
HTML format, in the `docs/md` and `public/docs` directories respectively.
Normally, Perl documentation is written in POD format, viewed cleanly on
MetaCPAN, but alas; this software is not published publicly. The POD format is
however available in code form, within the respective module files (there's
also a POD file for the `core.js` script, in the `docs/pod` directory).

##### [BCLC::Numbers](http://24.67.48.97:5000/docs/BCLC-Numbers.html)

This module houses the core REST API backend, and performs all of the
calculations and backend validation for the project.

##### [BCLC::Numbers::CSV](http://24.67.48.97:5000/docs/BCLC-Numbers-CSV.html)

The `CSV` module/class is responsible for reading in the historical CSV file,
and returning the aggregated data.

##### [BCLC::Numbers::DB](http://24.67.48.97:5000/docs/BCLC-Numbers-DB.html)

The `DB` module/class is responsible for performing the queries on the SQLite
database and returning the aggregated data.

##### [core.js](http://24.67.48.97:5000/docs/core.js.html)

This is the front-end, user-side script that performs all of the necessary
dynamic rendering of the UI, as well as provides live-time validation and
error/warning reporting directly to the user.

## Test Results

A test result summary can be found [here](http://24.67.48.97:5000/docs/test/test_summary.txt)
along with the complete [test results](http://24.67.48.97:5000/docs/test/test_results_full.txt).

REQ59847, IS27 Full Stack Developer
Due date: May 15, 2019 at 1:00 pm Pacitic Time
For questions, please email: Dea.DeJarisco@gov.bc.ca


Changes
Revision history for BCLC-Numbers

0.01    Date/time
        First version, released on an unsuspecting world.



Makefile.PL
use 5.006;
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME             => 'BCLC::Numbers',
    AUTHOR           => q{Steve Bertrand <steve.bertrand@gmail.com>},
    VERSION_FROM     => 'lib/BCLC/Numbers.pm',
    ABSTRACT_FROM    => 'lib/BCLC/Numbers.pm',
    LICENSE          => 'artistic_2',
    PL_FILES         => {},
    MIN_PERL_VERSION => '5.006',
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => '0',
    },
    BUILD_REQUIRES => {
        'Test::More' => '0',
    },
    PREREQ_PM => {
        Carp                     => 0,
        DBI                      => 0,
        Dancer2                  => 0,
        'Dancer2::Core::Request' => 0,
        JSON                     => 0,
        'Number::Format'         => 0,

        #'ABC'              => '1.6',
        #'Foo::Bar::Module' => '5.0401',
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'BCLC-Numbers-*' },
);


public/js/core.js
"use strict";

var lotto_nums = {};

var base_warning;
var populate_warning;
var no_results_warning;
var help;
var help_info;

var results_table;
var totals_table;

var display_all;

$(document).ready(function(){

    base_warning = $("#base_warning");
    populate_warning = $("#populate_warning");
    no_results_warning = $("#no_results_warning");
    help = $("#help");
    help_info = $("#help_info");

    results_table = $("#results_table");
    totals_table = $("#totals_table");

    display_all = $("#display_all");

    // hide all warnings and tables

    base_warning.hide();
    populate_warning.hide();
    no_results_warning.hide();
    help_info.hide();

    results_table.hide();
    totals_table.hide();

    $(".number_field").on("change input", function(){
        populate_warning.hide();

        var number = $(this).val();

        if (validate_number($(this).attr("id"), number)){
            $(this).removeClass("invalid").addClass("valid");
            base_warning.hide();
        }
        else {
            $(this).removeClass("valid").addClass("invalid");
            base_warning.show();
        }

        display_data();
    });

    display_all.on("change", function(){
        display_data();
    });

    help.on("change", function(){
        help.prop("checked") ? help_info.show() : help_info.hide();
    })
});

function display_data () {

    if (Object.keys(lotto_nums).length === 6) {

        var field_name_portion = "num";

        for (var i=1; i<7; i++){
            var field_id = "#" + field_name_portion + i;

            if ($(field_id).hasClass("invalid")){
                results_table.hide();
                totals_table.hide();
                return false;
            }
        }

        results_table.find("tr:gt(0)").remove();

        var fetch_data_params = {
            "numbers":      Object.values(lotto_nums),
            "display_all":  display_all.prop("checked"),
            "csv_source":   $("#data_source").val() === "csv"
        };

        $.ajax({
            async: true,
            type: 'GET',
            url: '/fetch_data/' + JSON.stringify(fetch_data_params),

            success: function (data) {
                var json = $.parseJSON(data);

                var winning_draws = json["winning_draws"];
                var total_spent_on_tickets = json["total_spent_on_tickets"];
                var total_number_payout = json["total_number_payout"];
                var net_win_loss = json["net_win_loss"];

                var error = json["error"];
                var error_msg = json["error_msg"];

                var backend_error_message = $("#backend_error_message");

                if (error){
                    $("#error_message").text(error_msg);
                    backend_error_message.show()
                }
                else {
                    backend_error_message.hide();
                }

                jQuery.each(winning_draws, function (index, draw) {
                    $("#results_table tr:last").after(
                        "<tr>" +
                        "<td>" + draw["DRAW DATE"] + "</td>" +
                        "<td>" + draw["WIN_AMOUNT"] + "</td>" +
                        "<td>" + draw["DRAW NUMBER"] + "</td>" +
                        "<td>" + draw["NUMBER_MATCHES"] + "</td>" +
                        "<td>" + draw["BONUS_MATCH"] + "</td>" +
                        "</tr>"
                    );
                });

                if (winning_draws.length > 0) {
                    no_results_warning.hide();
                    results_table.show();
                }
                else {
                    no_results_warning.show();
                }

                $("#total_won").text(total_number_payout);
                $("#total_spent_on_tickets").text(total_spent_on_tickets);
                $("#net_win_loss").text(net_win_loss);

                totals_table.show();
            }
        });
    }
}

function validate_number(id, num){
    var within_range = (num <= 49 && num >= 1);
    var duplicate = false;

    lotto_nums[id] = null;

    $.each(lotto_nums, function(k, v){
        if (v === num){
            duplicate = true;
        }
    });

    if (within_range && ! duplicate){
        lotto_nums[id] = num;
        return true;
    }

    return false;
}


lib/BCLC/Numbers/DB.pm
package BCLC::Numbers::DB;

use warnings;
use strict;

use Carp qw(croak);
use DBI qw(:sql_types);

our $VERSION = '0.01';

sub new {
    my ($class, $db_file) = @_;

    my $self = bless {}, $class;

    if (! defined $db_file){
        croak "new() requires a database file name to be sent in...";
    }

    $self->{db} = DBI->connect(
        "dbi:SQLite:dbname=$db_file",
        '',
        '',
        {
            #sqlite_use_immediate_transaction => 1,
            #sqlite_see_if_its_a_number => 1,
            RaiseError => $self->{db_err},
            AutoCommit => 1
        }
    ) or die $DBI::errstr;

    return $self;
}

sub retrieve {
    my ($self, %args) = @_;

    my $table = $args{table};

    if (! defined $table){
        croak "retrieve() requires a table name sent in...";
    }

    my $sth = $self->_db->prepare("SELECT * FROM $table");

    $sth->execute;

    return $sth->fetchall_arrayref({});
}

sub _db {
    return $_[0]->{db};
}
1;

__END__

=head1 NAME

BCLC::Numbers::DB - Provides database backend support for L<BCLC::Numbers>
L<Dancer2> web application

=head1 DESCRIPTION

Object Oriented module that provides database access for L<BCLC::Numbers>

=head1 METHODS

=head2 new($db_file)

Instantiates and returns a new L<BCLC::Numbers::DB> object, ready to process
requests.

Parameters:

    $db_file

Mandatory, String: The path and file name of the SQLite database file.

Return: An object of class L<BCLC::Numbers::DB>

=head2 retrieve(%args)

Compiles the entire Lotto 649 historical draw information.

Parameters:

All parameters are sent in as a hash.

    table => 'table_name'

Mandatory, String: The name of the database table to operate on.

Return: An Array Reference of Hash References, where each hash ref contains
the dataset's column name as the key, and its associated value as the value.

=head1 PRIVATE METHODS

=head2 _db

Returns the SQLite database connected object stored within the object.

Takes no parameters.

=head1 AUTHOR

Steve Bertrand, C<< <steve.bertrand at gmail.com> >>

=head1 LICENSE AND COPYRIGHT

Copyright 2019 Steve Bertrand.

Released under the Apache License, Version 2



lib/BCLC/Numbers/CSV.pm
package BCLC::Numbers::CSV;

use warnings;
use strict;

use Carp qw(croak);

our $VERSION = '0.01';

sub new {
    my ($class, $csv_file) = @_;

    if (! defined $csv_file){
        croak "new() requires a CSV file name to be sent in...";
    }

    my $self = bless {csv => $csv_file}, $class;

    return $self;
}

sub retrieve {
    my ($self) = @_;

    open my $csv_fh, '<', $self->{csv}
      or die "can't open CSV file $self->{csv}: $!";

    my @draws;
    my @headings;

    while (my $line = <$csv_fh>){
        chomp $line;

        if ($. == 1){
            @headings = split /,/, $line;
            next;
        }

        my %draw;

        @draw{@headings} = split /,/, $line;

        push @draws, \%draw;
    }

    return \@draws;
}

1;

__END__

=head1 NAME

BCLC::Numbers::CSV - Provides CSV file backend support for L<BCLC::Numbers>
L<Dancer2> web application

=head1 DESCRIPTION

Object Oriented module that provides CSV file access for L<BCLC::Numbers>

=head1 METHODS

=head2 new($csv_file)

Instantiates and returns a new L<BCLC::Numbers::CSV> object, ready to process
requests.

Parameters:

    $csv_file

Mandatory, String: The path and file name of the CSV file.

Return: An object of class L<BCLC::Numbers::CSV>

=head2 retrieve

Compiles the entire Lotto 649 historical draw information, using the CSV file.

Takes no parameters.

Return: An Array Reference of Hash References, where each hash ref contains
the dataset's column name as the key, and its associated value as the value.

=head1 AUTHOR

Steve Bertrand, C<< <steve.bertrand at gmail.com> >>

=head1 LICENSE AND COPYRIGHT

Copyright 2019 Steve Bertrand.

Released under the Apache License, Version 2


lib/BCLC/Numbers.pm
package BCLC::Numbers;

use warnings;
use strict;

our $VERSION = '0.01';

use BCLC::Numbers::CSV;
use BCLC::Numbers::DB;
use Dancer2;
use Dancer2::Core::Request;
use Number::Format qw(:subs :vars);

use constant {
    MAX_DRAW_NUMBER       => 3620,
    VALID_SEQUENCE_NUMBER => => 0,
};

my $db_file = 'data/649.db';
my $csv_file = 'data/649.csv';

my $csv = BCLC::Numbers::CSV->new($csv_file);
my $db = BCLC::Numbers::DB->new($db_file);

my $payout_table = payout_table();

get '/' => sub {
    return template 'main';
};

get '/fetch_data/:params' => sub {
    my $params = from_json param("params");

    my $numbers = $params->{numbers};
    my $display_all = $params->{display_all};
    my $csv_source = $params->{csv_source};

    return to_json fetch_data($numbers, $display_all, $csv_source);
};

sub convert_to_dollar {
    my ($int) = @_;
    return format_price($int //= 0, 2, '$');
}

sub fetch_data {
    my ($player_numbers, $display_all, $csv_source) = @_;

    $csv_source //= 1;

    my $draws;

    if ($csv_source){
        $draws = $csv->retrieve;
    }
    else {
        $draws = $db->retrieve(table => 'historical');
    }

    my @valid_draws;

    for my $draw (@$draws){

        next if filter($draw);

        my %winning_numbers;
        my $bonus_number;

        for (1..6){
            my $column = "NUMBER DRAWN $_";
            $winning_numbers{$draw->{$column}} = 1;
        }

        $bonus_number = $draw->{'BONUS NUMBER'};

        my %validate_numbers;
        my $backend_error = 0;

        my $match_count = 0;
        my $match_bonus = 0;

        for my $player_number (@$player_numbers){
            next if ! defined $player_number;

            if (! $validate_numbers{$player_number}){
                $validate_numbers{$player_number} = 1;
            }
            else {
                $backend_error = 1;
            }

            $backend_error = 1 if $player_number > 49;

            if ($backend_error){
                return {
                    error     => 1,
                    error_msg =>
                        "One or more numbers are either out of range, " .
                        "or is a duplicate. Please hit the Reset " .
                        "and try again!",
                }
            }
            if ($winning_numbers{$player_number}){
                $match_count++;
            }
            $match_bonus++ if $player_number == $bonus_number;
        }

        $draw->{NUMBER_MATCHES} = $match_count;
        $draw->{BONUS_MATCH} = $match_bonus;

        if ($draw->{NUMBER_MATCHES}){
            $draw->{WIN_AMOUNT}
              = convert_to_dollar(_calculate_win_value($draw));
        }

        push @valid_draws, $draw;
    }

    my $aggregate_data = _compile_data(\@valid_draws, $display_all);

    $aggregate_data->{error} = 0;

    # tag the data with its source

    $aggregate_data->{data_source} = $csv_source ? 'CSV' : 'DB';

    return $aggregate_data;
}

sub filter {
    my ($draw) = @_;

    if ($draw->{'SEQUENCE NUMBER'} != VALID_SEQUENCE_NUMBER
        || $draw->{'DRAW NUMBER'} > MAX_DRAW_NUMBER){
        return 1;
    }

    return 0;
}

sub payout_table {
    return {
        '2'  => 3,
        '2+' => 5,
        '3'  => 10,
        '4'  => 85,
        '5'  => 3000,
        '5+' => 250000,
        '6'  => 5000000,
    };
}

sub ticket_price {
    my ($draw) = @_;

    my $draw_number = $draw->{'DRAW NUMBER'};

    return 1 if $draw_number <= 2124;
    return 2 if $draw_number <= 2989;
    return 3;
}

sub _calculate_win_value {
    my ($draw) = @_;

    my $payout_key = $draw->{NUMBER_MATCHES};

    if ($payout_key && ($payout_key == 2 || $payout_key == 5)){
        if ($draw->{BONUS_MATCH}){
            $payout_key .= '+';
        }
    }

    return $payout_table->{$payout_key};
}

sub _compile_data {
    my ($all_draws, $display_all) = @_;

    my @winning_draws;
    my $total_spent_on_tickets;
    my $total_number_payout;

    for my $draw (@$all_draws) {

        $total_spent_on_tickets += ticket_price($draw);

        next if $draw->{NUMBER_MATCHES} < 2;

        $total_number_payout += _draw_payout($draw);

        if ($display_all){
            push @winning_draws, $draw;
        }
        elsif ($draw->{NUMBER_MATCHES} >= 4){
            # $85+
            push @winning_draws, $draw;
        }
    }

    my $net_win_loss
      = convert_to_dollar($total_number_payout - $total_spent_on_tickets);

    my $data = {
        winning_draws     => \@winning_draws,
        total_spent_on_tickets => convert_to_dollar($total_spent_on_tickets),
        total_number_payout => convert_to_dollar($total_number_payout),
        net_win_loss => $net_win_loss,
    };

    return $data;
}

sub _draw_payout {
    my ($draw) = @_;
    return _calculate_win_value($draw);
}

1;

__END__

=head1 NAME

BCLC::Numbers - Calculate Lotto 649 historical win information

=head1 DESCRIPTION

Using historical data (CSV or SQLite database), calculate and display win
information based on six unique numbers entered via a web UI.

=head1 REST ROUTES

=head2 /

Landing page. Loads up the C<main> L<Template::Toolkit> template that provides
the number inputs and various options.

This route is a C<get> request.

Takes no parameters, has no return.

=head2 /fetch_data/:params

This route front-ends the core workload of the application. It fetches and
returns all calculated draw information based on the numbers provided.

Parameters:

All parameters are passed in as a JSON string, converted to a Perl hash.

    numbers => [1, 2, 3, 4, 5, 6]

Six unique numbers within the range 1-49. Will C<croak()> if any of the numbers
are out of range, or non-unique.

    display_all => bool

By default, only wins $85 or higher will be displayed in the UI. Set this to
true to have all wins regardless of the value displayed on-screen.

Default: false

    csv_source => bool

This software has the optional ability to read directly from the CSV file
provided by BCLC, or an SQLite database containing the same data. Set this param
to C<true> to use the CSV reader.

Default: false

=head1 FUNCTIONS

=head2 convert_to_dollar($int)

Using L<Number::Format> convert an integer into a properly formatted dollar
string (eg: C<3000> becomes C<$3,000.00>).

Parameters:

    $int

Mandatory, Integer (or float): The integer or float you want converted to a
dollar string.

Return: Formatted dollar string.

=head2 fetch_data($player_numbers, $display_all, $csv_source)

Fetches, compiles, filters and aggregates all draw data.

Parameters:

    $player_numbers

Mandatory, Array Reference: An array reference of the six unique lotto numbers
between C<1> and C<49>.

    $display_all

Optional, bool: Set this to true to have all wins returned. By default, we only
calculate wins C<$85> or more as valid. Defaults to C<true> (C<1>).

    $csv_source

Optional, bool: Set to C<true> (C<1>), and we'll read our draw information from
BCLC's Lotto 649 historical CSV file directly. Send in C<false> (C<0>) to
read from a pre-compiled SQLite database containing the contents of the CSV.
Defaults to C<true>.

Return: Hash Reference. If any errors occurred, you'll receive two key/value
pairs: C<error => 1, error_message => "string">. Otherwise you'll receive a
valid data-filled href ready for processing.

=head2 filter($draw)

Performs any/all necessary filtering of draws from the aggregate pool.

Parameters:

    $draw

Mandatory, hash reference: A Lotto 649 draw hash reference as returned within
the array reference from the C<retrieve()> function from either L<BCLC::DB> or
L<BCLC::CSV>.

Return: C<1> (C<true>) if the draw sent in should be filtered out/skipped, and
C<0> (C<false>) if it's a valid draw.

=head2 payout_table

Returns a hash reference where the keys are the possible number match
combinations, and the values the dollar amount for each.

Takes no parameters.

Return: Hash Reference

=head2 ticket_price($draw)

Calculates the price of a ticket for an individual draw.

Parameters:

    $draw

Mandatory, hash reference: A Lotto 649 draw hash reference as returned within
the array reference from the C<retrieve()> function from either L<BCLC::DB> or
L<BCLC::CSV>.

Return, Integer: The cost of an individual draw's ticket.

=head1 PRIVATE FUNCTIONS

=head2 _calculate_win_value($draw)

Calculates the win value of a single draw.

Parameters:

    $draw

Mandatory, hash reference: A Lotto 649 draw hash reference as returned within
the array reference from the C<retrieve()> function from either L<BCLC::DB> or
L<BCLC::CSV>.

Return: An integer representing the dollar amount of the win for the draw
specified. Returns C<0> if no wins resulted.

=head2 _compile_data($all_draws, $display_all)

Compiles all the various data related to all draws.

Parameters:

    $all_draws

Mandatory, Array of Hashes: An array containing the full list of all draws as
returned by the C<retrieve()> method in L<BCLC::DB> or L<BCLC::CSV>.

    $display_all

Optional, bool: Send in a true value (eg C<1>) to have all wins compiled as
opposed to the default wins of $85 or higher. Defaults to false.

Return: A hash reference in the following format:

    winning_draws           => \@winning_draws,
    total_spent_on_tickets  => convert_to_dollar($total_spent_on_tickets),
    total_number_payout     => convert_to_dollar($total_number_payout),
    net_win_loss            => $net_win_loss,

=head2 _draw_payout($draw)

Returns the calculated payout of an individual draw.

Parameters:

    $draw

Mandatory, hash reference: A Lotto 649 draw hash reference as returned within
the array reference from the C<retrieve()> function from either L<BCLC::DB> or
L<BCLC::CSV>.

Return: Integer, the dollar amount won in the draw.

=head1 AUTHOR

Steve Bertrand, C<< <steve.bertrand at gmail.com> >>

=head1 LICENSE AND COPYRIGHT

Copyright 2019 Steve Bertrand.

Released under the Apache License, Version 2

